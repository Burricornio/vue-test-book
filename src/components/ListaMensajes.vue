
<template>
  <div>
    <header class="list-header">
      <slot name="header">
        Este es el header por defecto...
      </slot>
    </header>
    <ul class="lista-mensajes">
      <!-- <Mensaje
        v-for="(mensaje, indice) in mensajes"
        :key="indice"
        :mensaje= "mensaje"
        @mensaje-clickado="gestionarMensajeClick"
        >
      </Mensaje> -->
      <slot />
    </ul>
  </div>
</template>

<script>
// import Mensaje from './Mensaje.vue'

export default {
  name: 'list',
  // components: {
  //   Mensaje
  // },
  // props: ['mensajes'],
  methods: {
    // gestionarMensajeClick(mensaje) {
    //   console.log(mensaje)
    // }
  }
}
</script>
