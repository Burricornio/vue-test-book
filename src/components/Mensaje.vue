<template>
  <li style="margin-top: 10px" class="mensaje">{{ mensaje }}</li>
</template>

<script>
export default {
  props: ['mensaje'],
  created() {
    // Tengamos presente evitar la renderización anidada
    // cuando puede haber efectos secundarios (“side effects”),
    // dado que los life cycle hooks de los componentes hijos
    // (created, mounted) serán disparados y por lo tanto puede haber llamadas HTTP u otros efectos secundarios que no queremos que se ejecuten. Para probar lo que comentamos, puedes probar a añadir en el componente Mensaje.vue un console.log en el hook created():
    console.log('Creado')
  }
}
</script>
