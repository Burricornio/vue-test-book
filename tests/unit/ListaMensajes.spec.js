import { mount } from '@vue/test-utils'
import ListaMensajes from '@/components/ListaMensajes.vue'
import Mensaje from '@/components/Mensaje.vue'

describe('ListaMensajes.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(ListaMensajes, {
      propsData: {
        mensajes: ['Item test']
      }
    })
  })

  it('Tiene la estructura HTML esperada', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('ha recibido ["Item test"] como la propiedad mensajes', () => {
    expect(wrapper.vm.mensajes).toEqual(['Item test'])
  })

  // Usamos is para afirmar sobre el componente raíz y contains
  // para checkear si existen sub-componentes. Así como,
  // find, todos pueden recibir un Selector: ya sea un
  // Selector de CSS o un Componente

  it('Es un componente ListaMensajes', () => {
    expect(wrapper.is(ListaMensajes)).toBe(true)
    // O como Selector CSS del _markup_ generado:
    expect(wrapper.is('ul')).toBe(true)
  })

  it('Contiene un component <Mensaje>', () => {
    expect(wrapper.contains(Mensaje)).toBe(true)
    expect(wrapper.contains('.mensaje')).toBe(true)
  })

  it('Ambos ListaMensajes y Mensaje son instancias de Vue', () => {
    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.find(Mensaje).isVueInstance()).toBe(true)
  })

  // Testearemos la Estructuras en mayor detalle
  it('Elemento ensaje existe', () => {
    expect(wrapper.find('.mensaje').exists()).toBe(true)
  })

  it('Mensaje no es vacio', () => {
    expect(wrapper.find(Mensaje).isEmpty()).toBe(false)
  })

  it('Mensaje tiene un atributo de clase de valor "mensaje"', () => {
    expect(wrapper.find(Mensaje).classes()).toContain('mensaje')
  })

  it('Mensaje tiene el estilo "margin-top: 10"', () => {
  expect(wrapper.find(Mensaje).attributes().style).toBe('margin-top: 10px;')
  })

  // Eventos
  it('Llama a gestionarMensajeClick cuando @mensaje-click se produce', () => {
    const stub = jest.fn()
    wrapper.setMethods({ gestionarMensajeClick: stub })

    wrapper.find(Mensaje).vm.$emit('mensaje-clickado', 'gato')
    expect(stub).toBeCalledWith('gato')
  })
})
