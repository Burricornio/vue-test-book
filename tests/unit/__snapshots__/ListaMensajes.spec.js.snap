// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListaMensajes.vue Tiene la estructura HTML esperada 1`] = `
<div>
  <header
    class="list-header"
  >
    
      Este es el header por defecto...
    
  </header>
   
  <ul
    class="lista-mensajes"
  >
    <div
      class="mensaje-falso"
    />
  </ul>
</div>
`;
